### PR TITLE
hotfix(response-ratelimiting): handle header_value as table in parse_header

### LINKED ISSUE #2089: Duplicate HTTP Response Header for X-Kong-Limit cause error in response-ratelimiting plugin
### Summary

Duplicate HTTP Response Header for X-Kong-Limit cause error in response-ratelimiting plugin

### Steps To Reproduce
1. Enable the reponse-ratelimiting plugin
2. Have an upstream service return multiple header values for 'X-Kong-Plugin' e.g.
```
< Server: Apache-Coyote/1.1
< X-Application-Context: bootstrap:11500
< X-Kong-Limit: IDRequest=1
< X-Kong-Limit: IDAllocation=9
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Tue, 14 Feb 2017 16:14:50 GMT
```
3. You will get a `curl: (52) Empty reply from server`
4. The error.log of kong contains the following
```
/usr/local/share/lua/5.1/pl/stringx.lua:165: in function 'split'
...5.1/kong/plugins/response-ratelimiting/header_filter.lua:17: in function 'parse_header'
...5.1/kong/plugins/response-ratelimiting/header_filter.lua:39: in function 'execute'
...e/lua/5.1/kong/plugins/response-ratelimiting/handler.lua:21: in function 'header_filter'

```
5. According to https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2 having multiple headers with the same name is acceptable

### Additional Details & Logs

- Kong version 0.9.8
- Kong debug-level startup logs 
```
2017/02/14 15:55:03 [verbose] Kong: 0.9.8
2017/02/14 15:55:03 [debug] ngx_lua: 10006
2017/02/14 15:55:03 [debug] nginx: 1011002
2017/02/14 15:55:03 [debug] Lua: LuaJIT 2.1.0-beta2
2017/02/14 15:55:03 [debug] PRNG seed: 182151361912
2017/02/14 15:55:03 [verbose] no config file found at /etc/kong.conf
2017/02/14 15:55:03 [verbose] no config file found at /etc/kong/kong.conf
2017/02/14 15:55:03 [verbose] no config file, skipping loading
2017/02/14 15:55:03 [debug] KONG_CASSANDRA_CONTACT_POINTS ENV found with "kong-database"
2017/02/14 15:55:03 [debug] KONG_PG_HOST ENV found with "kong-database"
2017/02/14 15:55:03 [debug] KONG_DATABASE ENV found with "postgres "
2017/02/14 15:55:03 [debug] admin_listen = "0.0.0.0:8001"
2017/02/14 15:55:03 [debug] anonymous_reports = true
2017/02/14 15:55:03 [debug] cassandra_consistency = "ONE"
2017/02/14 15:55:03 [debug] cassandra_contact_points = {"kong-database"}
2017/02/14 15:55:03 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2017/02/14 15:55:03 [debug] cassandra_keyspace = "kong"
2017/02/14 15:55:03 [debug] cassandra_port = 9042
2017/02/14 15:55:03 [debug] cassandra_repl_factor = 1
2017/02/14 15:55:03 [debug] cassandra_repl_strategy = "SimpleStrategy"
2017/02/14 15:55:03 [debug] cassandra_ssl = false
2017/02/14 15:55:03 [debug] cassandra_ssl_verify = false
2017/02/14 15:55:03 [debug] cassandra_timeout = 5000
2017/02/14 15:55:03 [debug] cassandra_username = "kong"
2017/02/14 15:55:03 [debug] cluster_listen = "0.0.0.0:7946"
2017/02/14 15:55:03 [debug] cluster_listen_rpc = "127.0.0.1:7373"
2017/02/14 15:55:03 [debug] cluster_profile = "wan"
2017/02/14 15:55:03 [debug] cluster_ttl_on_failure = 3600
2017/02/14 15:55:03 [debug] custom_plugins = {}
2017/02/14 15:55:03 [debug] database = "postgres"
2017/02/14 15:55:03 [debug] dnsmasq = true
2017/02/14 15:55:03 [debug] dnsmasq_port = 8053
2017/02/14 15:55:03 [debug] log_level = "notice"
2017/02/14 15:55:03 [debug] lua_code_cache = "on"
2017/02/14 15:55:03 [debug] lua_package_cpath = ""
2017/02/14 15:55:03 [debug] lua_package_path = "?/init.lua;./kong/?.lua"
2017/02/14 15:55:03 [debug] lua_ssl_verify_depth = 1
2017/02/14 15:55:03 [debug] mem_cache_size = "128m"
2017/02/14 15:55:03 [debug] nginx_daemon = "on"
2017/02/14 15:55:03 [debug] nginx_optimizations = true
2017/02/14 15:55:03 [debug] nginx_worker_processes = "auto"
2017/02/14 15:55:03 [debug] pg_database = "kong"
2017/02/14 15:55:03 [debug] pg_host = "kong-database"
2017/02/14 15:55:03 [debug] pg_port = 5432
2017/02/14 15:55:03 [debug] pg_ssl = false
2017/02/14 15:55:03 [debug] pg_ssl_verify = false
2017/02/14 15:55:03 [debug] pg_user = "kong"
2017/02/14 15:55:03 [debug] prefix = "/usr/local/kong/"
2017/02/14 15:55:03 [debug] proxy_listen = "0.0.0.0:8000"
2017/02/14 15:55:03 [debug] proxy_listen_ssl = "0.0.0.0:8443"
2017/02/14 15:55:03 [debug] serf_path = "serf"
2017/02/14 15:55:03 [debug] ssl = true
2017/02/14 15:55:03 [verbose] prefix in use: /usr/local/kong
2017/02/14 15:55:03 [verbose] preparing nginx prefix directory at /usr/local/kong
2017/02/14 15:55:03 [verbose] saving serf identifier to /usr/local/kong/serf/serf.id
2017/02/14 15:55:03 [debug] searching for OpenResty 'resty' executable
2017/02/14 15:55:03 [debug] /usr/local/openresty/bin/resty -V: 'nginx version: openresty/1.11.2.1'
2017/02/14 15:55:03 [debug] found OpenResty 'resty' executable at /usr/local/openresty/bin/resty
2017/02/14 15:55:03 [verbose] saving serf shell script handler to /usr/local/kong/serf/serf_event.sh
2017/02/14 15:55:03 [verbose] SSL enabled, no custom certificate set: using default certificate
2017/02/14 15:55:03 [verbose] default SSL certificate found at /usr/local/kong/ssl/kong-default.crt
2017/02/14 15:55:03 [verbose] running datastore migrations
2017/02/14 15:55:03 [verbose] migrations up to date
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/dnsmasq.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/dnsmasq.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] dnsmasq already running at /usr/local/kong/pids/dnsmasq.pid
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/serf.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/serf.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] serf agent already running at /usr/local/kong/pids/serf.pid
2017/02/14 15:55:03 [debug] searching for OpenResty 'nginx' executable
2017/02/14 15:55:03 [debug] /usr/local/openresty/nginx/sbin/nginx -v: 'nginx version: openresty/1.11.2.1'
2017/02/14 15:55:03 [debug] found OpenResty 'nginx' executable at /usr/local/openresty/nginx/sbin/nginx
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] could not start Kong, stopping services
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] sending TERM signal to nginx running at /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2017/02/14 15:55:03 [debug] kill -s TERM `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2017/02/14 15:55:03 [verbose] leaving serf cluster
```

- Kong error logs 

```
2017/02/14 15:59:04 [error] 79#0: *15 failed to run header_filter_by_lua*: /usr/local/share/lua/5.1/pl/utils.lua:173: argument 1 expected a 'string', got a 'table'
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/pl/utils.lua:388: in function 'assert_arg'
        /usr/local/share/lua/5.1/pl/utils.lua:400: in function 'assert_string'
        /usr/local/share/lua/5.1/pl/utils.lua:173: in function 'usplit'
        /usr/local/share/lua/5.1/pl/stringx.lua:165: in function 'split'
        ...5.1/kong/plugins/response-ratelimiting/header_filter.lua:17: in function 'parse_header'
        ...5.1/kong/plugins/response-ratelimiting/header_filter.lua:39: in function 'execute'
        ...e/lua/5.1/kong/plugins/response-ratelimiting/handler.lua:21: in function 'header_filter'
        /usr/local/share/lua/5.1/kong.lua:174: in function 'header_filter'
        header_filter_by_lua:2: in function <header_filter_by_lua:1> while reading response header from upstream, client: 192.168.99.1, server: kong, request: "GET

```

- Kong configuration (registered APIs/Plugins & configuration file)
```
response-ratelimit plugin
```

- Operating System
```
Kong docker container
```

### PR DESCRIPTION
**Pull request by jdmonin**
### Summary

For issue #2089: If upstream sends multiple headers with the same name, combine their values per RFC 2616.

### Full changelog

* When parsing upstream response headers, handle header_value as table in parse_header

### Issues resolved

Fix #2089

### CODE REVIEW

**Comment by jdmonin** [comment]
I'd be happy to write a test for this also, but wasn't sure the best way to do the upstream-server part: I couldn't get httpbin or mockbin to return 2 headers with the same name. Would the test create a new lua thread that listens on a port and returns a set of hardcoded HTTP response headers?

File: kong/plugins/response-ratelimiting/header_filter.lua [hunk 1]
```diff
@@ -14,6 +14,14 @@ local _M = {}
...
   local increments = {}
   if header_value then
+    if (type(header_value) == "table") then
+      -- multiple headers with same name: combine per RFC 2616 section 4.2
+      local hval_str = header_value[1]
+      for i = 2,#header_value do
+        hval_str = hval_str .. "," .. header_value[i]
+      end
+      header_value = hval_str
+    end
```
Comment (thefosk) [inline kong/plugins/response-ratelimiting/header_filter.lua]: Hello @jdmonin, thx for the contribution. I believe that you changes can be replaced with a more concise version:

```lua
if type(header_value) == "table" then
  header_value = table.concat(header_value, ",")
end
```

And that would be it.

**Comment by jdmonin** [comment]
Thank you! I've amended and pushed the more concise version. I'm looking into the test now; thanks for the httpbin example.

File: kong/plugins/response-ratelimiting/header_filter.lua [hunk 2]
```diff
@@ -14,6 +14,9 @@ local _M = {}
 local function parse_header(header_value, limits)
   local increments = {}
   if header_value then
+    if type(header_value) == "table" then
+      header_value = table.concat(header_value, ",") -- combine per RFC 2616 section 4.2
+    end
```
Comment (thibaultcha) [inline kong/plugins/response-ratelimiting/header_filter.lua]: It really is too bad that we are concatenating this string with comma, only to be split again on the next instruction... How about not creating that string in the first place if not needed?

Comment (thefosk) [inline kong/plugins/response-ratelimiting/header_filter.lua]: @jdmonin can you please update?

**Comment by jdmonin** [comment]
Thanks @thefosk and @thibaultcha, that's definitely more elegant.  I've amended that and added tests for multiple x-kong-limit headers from upstream.

**Comment by jdmonin** [comment]
Thanks @thibaultcha for the review; I've pushed an update to address those points.

@thefosk I'm worried I might have misunderstood where you wanted the direct TCP connection to httpbin, if the test shouldn't use it and kong's own http client doesn't need it (since it can already supply a header value table to the plugin). Is my test as written sufficient?

**Comment by thibaultcha** [comment]
Manually merged to master and release/0.10.0. Thanks!

**Comment by jdmonin** [comment]
Glad to help, thank you very much!

**Pull request closed without merging**